_posts/2025-01-16-talking-tock-58.md

```diff
@@ -0,0 +1,42 @@
+---
+title: Talking Tock 58
+subtitle: New pointer type in the kernel
+authors: bradjc
+---
+
+Correctly expressing data types in Rust between userspace and the kernel with
+correctly captured semantics is challenging as different hardware platforms
+provide different data representations. The common example of this is 32-bit
+platforms versus 64-bit platforms which store different sized types in registers
+and therefore provide a different amount of data between userspace and the
+kernel. Emerging platforms, for example hardware with CHERI support, complicate
+this further. With [pull request #4174](https://github.com/tock/tock/pull/4174)
+merged, Tock has taken a step to improve this by adding `CapabilityPtr`, a type
+within the kernel to clearly express when data within the kernel is a pointer.
+
+The `CapabilityPtr` Type
+------------------------
+
+Fundamentally this type stores a pointer to memory. The name "capability"
+signifies this type also captures the validity of that pointer and whether the
+holder of the pointer can actually use the pointer to access memory.
+
+Traditionally, a pointer in `unsafe` Rust always has the capability to be
+dereferenced. That is, the hardware will access the referenced memory.
```

Comment:
> Traditionally, a pointer in `unsafe` Rust always has the capability to be dereferenced.

As written, that statement isn't really true:

1. The pointer may not point at a valid address (such an access would trap at the hardware level)
1. The pointer may point at an address that the MPU/PMP/MMU/etc is blocking access to (hardware trap)
1. The pointer may not have provenance that is valid for the dereference (the language's AM does not allow access through the pointer)

`CapabilityPtr` just adds one more element to that list (the pointer's data can contain permissions for that pointer).

Reply:
I think "in Rust" is doing a lot of heavy lifting there, the memory access will be compiled into the program by the Rust compiler, and unconditionally executed at runtime. Now, whether or not the dereference succeeds in hardware, is (arguably) a different matter. Perhaps this depends a bit on the particular semantics of "dereference" the reader assumes.

I think I can add a qualifier to make this perhaps a bit more clear.

Reply:
I don't think that's exactly true either. Given that trying to access a pointer might be UB (if it's not aligned / pointing to an allocated object / so on...) that Rust does not guarantee anything, including that hardware will access the referenced memory.

A compliant implementation of Rust (like MIRI, for instance) might make extra checks and do something like jump to a panic handler and this would be well within the bounds of "undefined behavior".